homeassistant/components/repetier/sensor.py

```diff
@@ -180,12 +180,16 @@ def update(self):
             return
         job_name = data['job_name']
         start = data['start']
-        print_time = data['print_time']
+        end = data['print_time']
         from_start = data['from_start']
-        time_end = start + round(print_time, 0)
-        self._state = datetime.utcfromtimestamp(time_end).isoformat()
-        remaining = print_time - from_start
+        time_end = start + round(end, 0)
+        remaining = end - from_start
         remaining_secs = int(round(remaining, 0))
+        state = time.strftime('%H:%M:%S', time.gmtime(remaining_secs))
```

Comment:
The state should be absolute utc isoformat.

Reply:
But tis is not a timestamp. It is seconds remaining of the print job, formatted as hour:minute:seconds

Ie. a 10 minutes printjob would show 00:10:00 at start and count down from this.

Reply:
Yes. I mean we shouldn't use seconds remaining as state. We should use an absolute utc timestamp isoformatted as state. Please revert to what I had in my suggestion.

The user will be able to show this absolute time stamp as relative time in the lovelace GUI.

Reply:
But an absolute timestamp in remaining and elapsed is not what it should be. It is not a timestamp.

I'll test your statement with and see how it can be achived byt lovelace - haven't seen this opportunity when setting up any of my other ~300 entities.

Reply:
https://www.home-assistant.io/lovelace/entities/#format

Reply:
Okay - will test when I get home from job :)

Reply:
`type: entities
entities:
  - entity: sensor.anet_a8_3d_printer_remaining
    format: relative`

It is not changing anything.
The state is still the timestamp which is wrong.

![HA-Repetier](https://user-images.githubusercontent.com/14296463/56917502-f0081180-6abb-11e9-88b7-f198f801d697.JPG)

(Yes, just tried changing the remaining sensor over to your example, not the elapsed)

Reply:
Not sure if it's this, but when I compare with the demo sensor that works, your example doesn't have the timezone in the time string.

This is the demo timestampt:
`2019-01-18T16:49:29+01:00`

Reply:
![timestamp_demo](https://user-images.githubusercontent.com/3181692/56934771-bc45df80-6aed-11e9-9ea5-8a4fd9284586.png)

Reply:
Don't know - I have used the source you provided, and correct - the timestamp does not include the timezone.

```
class RepetierRemainingSensor(RepetierSensor):
    """Class to create and populate a Repetier Time Remaining Sensor."""

    def device_class(self):
        """Return the device class."""
        return DEVICE_CLASS_TIMESTAMP

    def update(self):
        """Update the sensor."""
        data = self._get_data()
        if data is None:
            return
        job_name = data['job_name']
        start = data['start']
        print_time = data['print_time']
        from_start = data['from_start']
        time_end = start + round(print_time, 0)
        self._state = datetime.utcfromtimestamp(time_end).isoformat()
        remaining = print_time - from_start
        remaining_secs = int(round(remaining, 0))
        _LOGGER.debug(
            "Job %s remaining %s",
            job_name, time.strftime('%H:%M:%S', time.gmtime(remaining_secs)))

```

Besides that, I see the following as a device class - in my eyes that doesn't look correct:

```
 "device_class": "<bound method RepetierElapsedSensor.device_class of <Entity Anet_A8_3D_Printer_elapsed: None>>",
  "friendly_name": "Anet_A8_3D_Printer_elapsed"
```

Reply:
Followed [this](https://stackoverflow.com/questions/19654578/python-utc-datetime-objects-iso-format-doesnt-include-z-zulu-or-zero-offset) to get the timezone added (which is non standard in Python according to the link)

It did absolutely no difference.

![HA-Repetier-Test-2](https://user-images.githubusercontent.com/14296463/56981610-91a46700-6b7f-11e9-8dd8-2e0bd222f4e7.JPG)

Can we agree that this is a non working solution?
Perhaps for a future version/update of this component, when the core devs get this sorted out, we can change to this solution - until then I see no other solution than the one I originally proposed, showing %H:%M:%S in state.

Reply:
Tracked the error to be a missing `@property` before the `def device_class`

Now it just returns `Invalid Date` in the frontend, when the sensor is unavailable.

Is it okay to dynamically change the device_class during runtime?
Ie. when the sensor is unavailable then make it return None and when the sensor is available make it return DEVICE_CLASS_TIMESTAMP

Reply:
Good that you found my bug.

No it's not ok to change device class dynamically. The frontend probably needs better handling of the `None` state.

Reply:
Okay - so the suggestion would be to leave it with `Invalid date` for now?

Then I just need to figure out the part with dynamically adding sensors.
I'll be going to the states next week for a conference, so don't expect too much new stuff from sunday and a week ahead ;)

I'll be using the next few days, up until I leave, to try and figure out the callbacks. If you know of a simpler example than MySensors, please speak up :)